stock_orderpoint_generator/models/product.py

```diff
@@ -0,0 +1,122 @@
+        record = super().create(vals)
+        if vals.get('auto_orderpoint_template_ids'):
+            record.auto_orderpoint_template_ids.create_orderpoints(record)
+        return record
+
+    @api.multi
+    def write(self, vals):
+        result = super().write(vals)
+        if vals.get('auto_orderpoint_template_ids'):
+            orderpoint_templates = self.mapped('auto_orderpoint_template_ids')
+            orderpoint_templates.create_orderpoints(self)
+        return result
+
+    def _compute_historic_quantities_dict(
+            self, location_id=False, from_date=False, to_date=False):
+        """Returns a dict of products with a dict of historic moves as for
+           a list of historic stock values resulting from those moves. If
+           a location_id is passed, we can restrict it to such location"""
+        location = location_id and location_id.id
+        domain_quant_loc, domain_move_in_loc, domain_move_out_loc = (
+            self.with_context(location=location)._get_domain_locations())
+        if not to_date:
+            to_date = fields.Datetime.now()
+        domain_move_in = domain_move_out = ([
+            ('product_id', 'in', self.ids),
+            ('state', '=', 'done'),
+        ] + domain_move_in_loc)
+        domain_move_out = ([
+            ('product_id', 'in', self.ids),
+            ('state', '=', 'done'),
+        ] + domain_move_out_loc)
+        if from_date:
+            domain_move_in += [('date', '>=', from_date)]
+            domain_move_out += [('date', '>=', from_date)]
+        domain_move_in += [('date', '<=', to_date)]
+        domain_move_out += [('date', '<=', to_date)]
+        move_obj = self.env['stock.move']
+        # Positive moves
+        moves_in = move_obj.search_read(
+            domain_move_in, ['product_id', 'product_qty', 'date'],
+            order='date asc')
+        # We'll convert to negative these quantities to operate with them
+        # to obtain the stock snapshot in every moment
+        moves_out = move_obj.search_read(
+            domain_move_out, ['product_id', 'product_qty', 'date'],
+            order='date asc')
+        for move in moves_out:
+            move['product_qty'] *= -1
+        # Merge both results and group them by product id as key
+        moves = moves_in + moves_out
+        # Obtain a dict with the stock snapshot for the relative date_from
+        # otherwise, the first move will counted as first stock value. We
+        # default the compute the stock value anyway to default the value
+        # for products with no moves for the given period
+        initial_stock = {}
+        initial_stock = self.with_context(
+            location=location)._compute_quantities_dict(
+                False, False, False, to_date=from_date or to_date)
+        product_moves_dict = {}
+        for move in moves:
+            product_moves_dict.setdefault(move['product_id'][0], {})
+            product_moves_dict[move['product_id'][0]].update({
```

Comment:
Why don't you do it on the previous line?

Reply:
There could be no key yet